docs/design/sum_types.md

```diff
@@ -0,0 +1,197 @@
+  None
+}
+```
+
+This declares a sum type named `OptionalI32` with two alternatives: `Some`,
+which holds a single `i32` value, and `None`, which is empty. Choice types can
+also be parameterized, like class types:
+
+```carbon
+choice Optional(T:! Type) {
+  Some(value: T),
+  None
+}
+```
+
+An alternative declaration consists of the alternative name, followed by an
+optional parameter list in parentheses, much like the syntax of a function
+declaration. A value of a function-like alternative is specified by "calling" it
+like a function, and a value of a "bare" alternative like `None` is specified by
+naming it:
+
+```carbon
+var my_opt: Optional(i32) = Optional(i32).None;
+my_opt = Optional(i32).Some(42);
+```
+
+The value of a choice type can be inspected using a `match` statement:
+
+```carbon
+match (my_opt) {
+  case .Some(the_value: i32) => {
+    Print(the_value);
+  }
+  case .None => {
+    Print("None");
+  }
+}
+```
+
+## User-defined sum types
+
+`choice` declarations are a convenience shorthand for common use cases, but they
+have limited flexibility. There is no way to control the representation of a
+`choice` type, or define methods or other members for it (although you can
+extend it to implement interfaces, using
+[`external impl`](generics/overview.md#implementing-interfaces) or
+[`adapter`](generics/overview.md#adapting-types)). However, a `class` type can
+be extended to behave like a sum type. This is much more verbose than a `choice`
+declaration, but gives the author full control over the representation and class
+members.
+
+The ability to create instances of the sum type can be straightforwardly
+emulated with factory functions and static constants (although the latter
+[are not yet designed](classes.md#no-static-variables)), and the internal
+storage layout will presumably involve untagged unions or some other low-level
+storage primitive which hasn't been designed yet, but the key to defining a sum
+type's interface is enabling it to support pattern matching. To do that, the sum
+type has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+Here's how that would look if `Optional` were defined as a class:
+
+```carbon
+class Optional(T:! Type) {
+  // Factory functions
+  fn Some(value: T) -> Self;
+  fn None() -> Self;
+
+  private var has_value: bool;
+  private var value: T;
+
+  interface MatchContinuation {
+    let template ReturnType:! Type;
+    fn Some(value: T) -> ReturnType;
+    fn None() -> ReturnType;
+  }
+
+  external impl as Match(MatchContinuation) {
+    fn Op[me: Self, Continuation:! MatchContinuation](
+        continuation: Continuation*) {
```

Comment:
Is the `*` here needed? (I don't think continuation is being mutated)

If needed, should it be `addr continuation` instead?

Reply:
> nishchay.space

My mistake